projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,286 @@
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+}
+
+static StatusCode prv_read_register(const LTCAFESettings *afe,
+                                    LTCAFERegister reg, uint8_t *data, size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(reg_cmd, cmd);
+
+  prv_wakeup_idle(afe);
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+// read from a voltage register
+static StatusCode prv_read_voltage(LTCAFESettings *afe,
+                                    LTCAFEVoltageRegister reg, LTCAFEVoltageRegisterGroup *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  size_t len = sizeof(LTCAFEVoltageRegisterGroup) * PLUTUS_AFE_DEVICES_IN_CHAIN;
+  return prv_read_register(afe, s_voltage_reg[reg], (uint8_t *)data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_NOT_PERMITTED
+                  | LTC6804_CNVT_CELL_ALL | (mode << 7);
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(adcv, cmd);
+
+  prv_wakeup_idle(afe);
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(100);
+}
+
+static void prv_trigger_aux_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADAX
+  uint16_t adax = LTC6804_ADAX_RESERVED | LTC6804_ADAX_GPIO1 | (mode << 7);
+
+  uint8_t cmd[4] = { 0 };
+  prv_build_cmd(adax, cmd);
+
+  prv_wakeup_idle(afe);
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(10);
+}
+
+// write config to all devices
+static StatusCode prv_write_config(const LTCAFESettings *afe, uint8_t gpio_enable_pins) {
+  // see p.54 in datasheet
+  LTCAFEWriteConfigPacket config_packet = { 0 };
+
+  // WRCFG
+  config_packet.wrcfg.data_hi = (uint8_t)(LTC6804_WRCFG_RESERVED >> 8);
+  config_packet.wrcfg.data_lo = (uint8_t)(LTC6804_WRCFG_RESERVED & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate((uint8_t *)&config_packet.wrcfg, 2);
+  config_packet.wrcfg.pec_hi = (uint8_t)(cmd_pec >> 8);
+  config_packet.wrcfg.pec_lo = (uint8_t)(cmd_pec & 0xFF);
+
+  // send CFGR registers starting with the bottom slave in the stack
+  for (uint8_t device = PLUTUS_AFE_DEVICES_IN_CHAIN; device > 0; --device) {
+    uint8_t curr_device = PLUTUS_AFE_DEVICES_IN_CHAIN - device;
+    uint8_t enable = gpio_enable_pins;
+    uint16_t undervoltage = 0;
+    uint16_t overvoltage = 0;
+
+    uint16_t cell_index = curr_device * LTC6804_CELLS_PER_DEVICE;
+    config_packet.devices[curr_device].reg.discharge_c1 = s_discharging_cells[cell_index + 0];
```

Comment:
I can use a 12 bit value to represent this, which was what Titus was talking about earlier. If you feel really strongly about it, I'll make the change, but I felt this captured the intent better.